ipa-core/src/protocol/ipa_prf/malicious_security/verifier.rs

```diff
@@ -10,6 +10,43 @@ use crate::{
     },
 };
 
+/// This function computes the shares that sum to zero from the zero-knowledge proofs.
+pub fn compute_g_differences<F, const P: usize, const λ: usize>(
+    zkp: &Vec<[F; P]>,
+    challenges: &[F],
+) -> Vec<F>
+where
+    F: PrimeField,
+{
+    // compute denominator
+    let lagrange_denominator: CanonicalLagrangeDenominator<F, P> =
+        CanonicalLagrangeDenominator::<F, P>::new();
+
+    // compute g_r with "empty" first spot
+    let g_r = iter::once(F::ZERO)
+        .chain(
+            challenges
+                .iter()
+                .zip(zkp)
+                .map(|(challenge, zkp)| interpolate_at_r(zkp, *challenge, &lagrange_denominator)),
```

Comment:
I don't understand why we are calling interpolate at r and providing r=0 for the first proof. Am I misunderstanding what the code does? Because that's not secure.

Reply:
Here we start with a 0 and append all the g_r = zkp(r) values, i.e. (0, zkp_1(r1) , ..., zkp_n(rn)). For the g_sums we do the opposite, i.e. (gsum_1,...,gsum_n, 0) so we get the differences (gsum_1-0, gsum_2-zkp_1(r1),...,0-zkp_n(rn))